pkg/http/handler/timeout.go

```diff
@@ -57,60 +59,61 @@ func NewTimeoutHandler(h http.Handler, msg string, firstByteTimeout time.Duratio
 		body:             msg,
 		firstByteTimeout: firstByteTimeout,
 		idleTimeout:      idleTimeout,
+		clock:            clock.RealClock{},
 	}
 }
 
 func (h *timeoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	ctx, cancel := context.WithCancel(r.Context())
 	defer cancel()
 
-	// done is closed when h.handler.ServeHTTP completes and contains
-	// the panic from h.handler.ServeHTTP if h.handler.ServeHTTP panics.
-	done := make(chan interface{})
-	tw := &timeoutWriter{w: w}
-	go func() {
-		defer func() {
-			defer close(done)
-			if p := recover(); p != nil {
-				done <- p
-			}
-		}()
-		h.handler.ServeHTTP(tw, r.WithContext(ctx))
-	}()
-
-	firstByteTimeout := getTimer(h.firstByteTimeout)
+	firstByteTimeout := getTimer(h.clock, h.firstByteTimeout)
 	var firstByteTimeoutDrained bool
 	defer func() {
 		putTimer(firstByteTimeout, firstByteTimeoutDrained)
 	}()
 
-	var idleTimeout *time.Timer
+	var idleTimeout clock.Timer
 	var idleTimeoutDrained bool
 	if h.idleTimeout > 0 {
-		idleTimeout = getTimer(h.idleTimeout)
+		idleTimeout = getTimer(h.clock, h.idleTimeout)
 		defer func() {
 			putTimer(idleTimeout, idleTimeoutDrained)
 		}()
 	}
 	var idleTimeoutCh <-chan time.Time
 	if idleTimeout != nil {
-		idleTimeoutCh = idleTimeout.C
+		idleTimeoutCh = idleTimeout.C()
 	}
 
+	// done is closed when h.handler.ServeHTTP completes and contains
+	// the panic from h.handler.ServeHTTP if h.handler.ServeHTTP panics.
+	done := make(chan interface{})
+	tw := &timeoutWriter{w: w, clock: h.clock}
+	go func() {
+		defer func() {
+			defer close(done)
+			if p := recover(); p != nil {
+				done <- p
+			}
+		}()
+		h.handler.ServeHTTP(tw, r.WithContext(ctx))
+	}()
+
```

Comment:
Note: Had to move this block to guarantee that the timers above are setup before the handler is called.